### ISSUE TITLE
Align AST for TSEnumDeclaration with TSInterfaceDeclaration, TSModuleDeclaration, etc

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#1305]
For each `TSEnumDeclaration`, we place its members directly on this AST node.
```ts
export interface TSEnumDeclaration extends BaseNode {
  type: AST_NODE_TYPES.TSEnumDeclaration;
  id: Identifier;
  members: TSEnumMember[];
  const?: boolean;
  declare?: boolean;
  modifiers?: Modifier[];
  decorators?: Decorator[];
}
```

For each `TSInterfaceDeclaration`, we have a separate `TSInterfaceBody` node, upon which we place the members.
```ts
export interface TSInterfaceDeclaration extends BaseNode {
  type: AST_NODE_TYPES.TSInterfaceDeclaration;
  body: TSInterfaceBody;
  id: Identifier;
  typeParameters?: TSTypeParameterDeclaration;
  extends?: ExpressionWithTypeArguments[];
  implements?: ExpressionWithTypeArguments[];
  decorators?: Decorator[];
  abstract?: boolean;
  declare?: boolean;
}

export interface TSInterfaceBody extends BaseNode {
  type: AST_NODE_TYPES.TSInterfaceBody;
  body: TypeElement[];
}
```

I'm not sure why there's this inconsistency between the two, but I think we should consider aligning the two.

I'm not sure why `TSEnumDeclaration` was created without a body, but I think that creating a `TSEnumBody` node seems like the correct thing to do.
Doing so would align it with not only `TSInterfaceDeclaration`, but also `TSModuleDeclaration`, `ClassDeclaration` and `FunctionDeclaration`.

**Comment by armano2** [comment]
hi, @bradzacher reason behind not using `*body` node for enum was alignment between babel ast for typescript and typescript itself.
https://github.com/typescript-eslint/typescript-eslint/blob/835378e505f462d965ce35cc4c81f8eee1704a30/packages/typescript-estree/src/convert.ts#L2471-L2473

#### note
one thing to keep in mind while changing this `no-shadow`.
http://www.typescriptlang.org/play/?ssl=10&ssc=1&pln=1&pc=1#code/MYewdgzgLgBAZjAvDA7Gg3AKAKZgK4C2MAogIwwDemMN8SMpANDAPQswDKAFgIYAmIAO4QYULthh4ADlOwAnGAAM4igHTVaAc3pxmbeqUwBfHPiLEATJQ0045ZExsxtyXa3YAVLgEsRv0eKSMvJKKqr0aCjGmEA

**Comment by bradzacher** [comment]
> Enum is not behaving in exactly same way as object.

Sure it's not the same as an object, but in your example, it is equivalent to the following:

```ts
const f = 777;
function E1() {
    const f = 1; // Shadows the upper `f`.
    const g = f; // = 1
}
function E2() {
    const f1 = 1;
    const g = f; // This is the upper `f`. = 777
}
```

Which is kinda my point - it behaves more like a new block scope (see functions, or BlockStatements)

> alignment between babel ast for typescript and typescript itself

To me it seems like we should look to mirroring what ESTree does, instead of what babel and TS do. Babel and TS use their AST much differently than ESTree.

**Comment by JoshuaKGoldberg** [comment]
#8920 was merged into the `v8` branch. ✅
